milestones/data.py

```diff
@@ -0,0 +1,373 @@
+            name=relationship,
+            active=True
+        )
+    except internal.MilestoneRelationshipType.DoesNotExist:
+        if relationship in ['requires', 'fulfills']:
+            return internal.MilestoneRelationshipType.objects.create(
+                name=relationship,
+                active=True
+            )
+        else:
+            raise exceptions.InvalidMilestoneRelationshipTypeException()
+
+
+# PUBLIC METHODS
+def create_milestone(milestone):
+    """
+    Inserts a new milestone into app/local state
+    Returns a dictionary representation of the object
+    """
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    milestone, __ = internal.Milestone.objects.get_or_create(  # pylint: disable=invalid-name
+        namespace=milestone_obj.namespace,
+        name=milestone_obj.name,
+        active=True,
+        defaults={
+            'description': milestone_obj.description,
+        }
+    )
+    return serializers.serialize_milestone(milestone)
+
+
+def update_milestone(milestone):
+    """
+    Updates an existing milestone in app/local state
+    Returns a dictionary representation of the object
+    """
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    try:
+        milestone = internal.Milestone.objects.get(id=milestone_obj.id)
+        milestone.name = milestone_obj.name
+        milestone.namespace = milestone_obj.namespace
+        milestone.description = milestone_obj.description
+        milestone.active = milestone_obj.active
+    except internal.Milestone.DoesNotExist:
+        raise exceptions.InvalidMilestoneException()
+    return serializers.serialize_milestone(milestone)
+
+
+def delete_milestone(milestone):
+    """
+    Deletes an existing milestone from app/local state
+    No return currently defined for this operation
+    """
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    _delete_milestone(milestone_obj)
+
+
+def _delete_milestone(milestone):
+    """
+    Internal helper for milestone removals -- also removes defined dependencies
+    """
+    # Remove related entities, and then remove the Milestone
+    internal.CourseMilestone.objects.filter(
+        milestone=milestone.id).delete()
+    internal.CourseContentMilestone.objects.filter(
+        milestone=milestone.id).delete()
+    internal.UserMilestone.objects.filter(
+        milestone=milestone.id).delete()
+    internal.Milestone.objects.filter(
+        id=milestone.id).delete()
+
+
+def fetch_milestones(milestone):
+    """
+    Retrieves a set of matching milestones from app/local state
+    Returns a list-of-dicts representation of the object
+    """
+    if milestone is None:
+        raise exceptions.InvalidMilestoneException()
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    if milestone_obj.id is not None:
+        return serializers.serialize_milestones(internal.Milestone.objects.filter(
```

Comment:
I'll need to look at your serializers and see if you are referencing related tables in the serializers. If so, this may cause some unnecessary round_trips to the database during the serialization phase. If so, you might want to call select_related() [I think that's the name of the ORM to do the JOINs on the initial SQL query]